Add documentation on follower reads

rmloveland opened the pull request:
Addresses #3287

Summary of changes:

- Update transaction layer docs where reads from the leaseholder are
  mentioned as a thing; add a little section noting that follower reads
  are a thing as well.

- Add a new page 'Follower Reads' where we go into more detail about
  what they are and how they work, and some of the use cases
  they (will?) enable.

- Update sidenav (2.1 only)

- Add ASCII diagrams used to generate the PNGs with DITAA for
  safekeeping

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/3288)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/2b1d0ca373a9ab1b1b560cece11009429199e30d/

rmloveland commented:
Hi @tschottdorf!  In addition to your general review of the doc
for technical correctness and understandabilty/clarity (and
whatever other criteria you want to review on, all of which I will
appreciate), I had a few questions:

1.  Would it be accurate to add something like "follower reads are
    an optimization that happens automatically in the background,
    there is no user-facing changes needed to your application or
    SQL code to take advantage of the improvements" or something
    similar?

2.  Should we remove mention of the analytics and cluster recovery
    use cases at this time?  In other words, are those things we
    might have later that are enabled by this feature, but that we
    don't necessarily have now and might not for a while?

3.  Although it is mentioned a bunch in the RFC, I have tried to
    avoid any mention of quiescent Raft in describing follower
    reads.  I spoke to Alex in Core office hours the other day and
    confirmed my understanding from the [stability blog post](https://www.cockroachlabs.com/blog/cockroachdb-stability-from-1-node-to-100-nodes/) that
    node quiescence is basically a Raft optimization but not
    something users should have to care about.  My guess was that
    therefore we don't necessarily want to document it, which Alex
    agreed with.  Furthermore, quiescence and related things are
    not mentioned anywhere in our docs currently (that I could
    find). For those reasons, I've tried to keep mentions of
    quiescence out of this doc.  Do you agree with that approach?

4.  I'm a bit worried that this doc is "not even wrong" since it's
    based on me reading the RFC(s) and various PR comments - if so
    let's discuss how to address.

5.  All of the above is with the understanding that these docs will
    not get merged until
    <https://github.com/cockroachdb/cockroach/pull/21056> is in

rmloveland commented:
Direct links are:

- [Transaction layer - new section](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/2b1d0ca373a9ab1b1b560cece11009429199e30d/dev/architecture/transaction-layer.html#time-and-hybrid-logical-clocks)

- [Follower Reads page](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/2b1d0ca373a9ab1b1b560cece11009429199e30d/dev/follower-reads.html) (possibly a new page is not best approach, happy to discuss)

tschottdorf commented:
I think the doc is correct, but too technical :-)

1. absolutely! The core doc addendum here I think would be near `AS OF SYSTEM TIME` to explain that these queries can (usually) lean on all of the followers to serve reads, along with promoting `AS OF SYSTEM TIME` usage suitable in other places in the docs (perhaps the SELECT docs, or general performance tuning tips).
2. yep
3. yep, I'd leave out most of the technicalities.
4. it's correct, just, as I mentioned above, too technical, I think.
5. :thumbsup:  I think it'll be quite a while until we should merge these docs. Follower reads may even still slip! We're not more than 25% of the way.

---

Reviewed 9 of 9 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/follower-reads.md, line 7 at r1](https://reviewable.io/reviews/cockroachdb/docs/3288#-LFqnHac2GfVCLiDiCQX:-LFqnHac2GfVCLiDiCQY:b-3nmmc3) ([raw file](https://github.com/cockroachdb/docs/blob/2b1d0ca373a9ab1b1b560cece11009429199e30d/v2.1/follower-reads.md#L7)):*
> ```Markdown
> ---
> 
> Follower reads are [consistent](strong-consistency.html) reads at historical timestamps from follower [replicas](architecture/overview.html#terms). They make the non-leaseholder replicas in a range suitable sources for historical reads. This can increase read throughput by allowing every replica to serve read requests. For geo-distributed clusters, this can translate to significant latency improvements.
> ```

Clarify what "historical timestamps" means: follower reads will only be used for queries using `AS OF SYSTEM TIME` (at least initially, there's no fundamental reason not to use them for regular queries, but it's unlikely that many regular queries will be able to use them).

---

*[v2.1/follower-reads.md, line 9 at r1](https://reviewable.io/reviews/cockroachdb/docs/3288#-LFqndEC9Sag_RGze7c6:-LFqndEC9Sag_RGze7c7:bj8uhqz) ([raw file](https://github.com/cockroachdb/docs/blob/2b1d0ca373a9ab1b1b560cece11009429199e30d/v2.1/follower-reads.md#L9)):*
> ```Markdown
> Follower reads are [consistent](strong-consistency.html) reads at historical timestamps from follower [replicas](architecture/overview.html#terms). They make the non-leaseholder replicas in a range suitable sources for historical reads. This can increase read throughput by allowing every replica to serve read requests. For geo-distributed clusters, this can translate to significant latency improvements.
> 
> In addition to optimizing certain types of reads, follower reads enable use cases such as:
> ```

These are possible future features - I wouldn't want to go out and claim that this is what they enable.

---

*[v2.1/follower-reads.md, line 18 at r1](https://reviewable.io/reviews/cockroachdb/docs/3288#-LFqnizr4QnnAl3R87mu:-LFqnizr4QnnAl3R87mv:bfv0pxa) ([raw file](https://github.com/cockroachdb/docs/blob/2b1d0ca373a9ab1b1b560cece11009429199e30d/v2.1/follower-reads.md#L18)):*
> ```Markdown
> 
> {{site.data.alerts.callout_info}}
> Follower reads are only supported on ranges with epoch-based leases. In other words, they only work on ranges that make up non-system tables.
> ```

I don't think people should have to care about this. They will work for all of their data, after all.

---

*[v2.1/architecture/transaction-layer.md, line 81 at r1](https://reviewable.io/reviews/cockroachdb/docs/3288#-LFqnygD4K8mTVhL39v9:-LFqnygD4K8mTVhL39vA:b-y08i6p) ([raw file](https://github.com/cockroachdb/docs/blob/2b1d0ca373a9ab1b1b560cece11009429199e30d/v2.1/architecture/transaction-layer.md#L81)):*
> ```Markdown
> In addition to the case described above, where the leaseholder for a range is responsible for serving reads, there are cases in which non-leaseholder replicas can serve reads for data stored on a range, provided certain criteria are met.  Such cases are called _follower reads_.
> 
> Follower reads are [consistent](../strong-consistency.html) reads at historical timestamps from follower replicas.  They make the non-leaseholder replicas in a range suitable sources for historical reads.  This can increase read throughput by allowing every replica to serve read requests.  For geo-distributed clusters, this can translate to significant latency improvements.
> ```

I would simplify this to say that "for `AS OF SYSTEM TIME` queries with timestamps suitably in the past (note: we should define what "suitably" means, by default 10-20s should do), under normal operation followers will be able to ....

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3288#-:-LFqoDoD3E3hBOCrRIpc:bx2chyd)*
<!-- Sent from Reviewable.io -->

rmloveland commented:
Thanks for the review @tschottdorf!  I agree the follower reads page is too technical with all the diagrams and stuff.  I guess that was part of a learning journey for me that we don't need to inflict on the majority of our users.

For now I have only made updates to address your specific comments and to add that "follower reads are
an optimization that happens automatically in the background etc.", since it sounds like we should put this PR on ice for a bit while the feature is actually developed.  As you note there will be a bit more work to do to weave it into everything else (AOST, SELECT, etc.) once it is closer to shipping. :-) 

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/follower-reads.md, line 7 at r1](https://reviewable.io/reviews/cockroachdb/docs/3288#-LFqnHac2GfVCLiDiCQX:-LG74DJ97RcY-YDTo-Ls:b-fqcs5x) ([raw file](https://github.com/cockroachdb/docs/blob/2b1d0ca373a9ab1b1b560cece11009429199e30d/v2.1/follower-reads.md#L7)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Clarify what "historical timestamps" means: follower reads will only be used for queries using `AS OF SYSTEM TIME` (at least initially, there's no fundamental reason not to use them for regular queries, but it's unlikely that many regular queries will be able to use them).
</blockquote></details>

Updated to add the phrase "By 'historical timestamps', we mean that follower reads will generally only be used for queries using [`AS OF SYSTEM TIME`](as-of-system-time.html) with timestamps suitably far in the past (10-20 seconds)."

---

*[v2.1/follower-reads.md, line 9 at r1](https://reviewable.io/reviews/cockroachdb/docs/3288#-LFqndEC9Sag_RGze7c6:-LG74B1HEeS7wcZfopFu:b-70ci5x) ([raw file](https://github.com/cockroachdb/docs/blob/2b1d0ca373a9ab1b1b560cece11009429199e30d/v2.1/follower-reads.md#L9)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

These are possible future features - I wouldn't want to go out and claim that this is what they enable.
</blockquote></details>

Fixed by deletion, thanks.

---

*[v2.1/follower-reads.md, line 18 at r1](https://reviewable.io/reviews/cockroachdb/docs/3288#-LFqnizr4QnnAl3R87mu:-LG749vm1OpNIEKkc5Ws:b-bzsifs) ([raw file](https://github.com/cockroachdb/docs/blob/2b1d0ca373a9ab1b1b560cece11009429199e30d/v2.1/follower-reads.md#L18)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I don't think people should have to care about this. They will work for all of their data, after all.
</blockquote></details>

Fixed by deletion.

---

*[v2.1/architecture/transaction-layer.md, line 81 at r1](https://reviewable.io/reviews/cockroachdb/docs/3288#-LFqnygD4K8mTVhL39v9:-LG746p5AuS4oISYK5h9:b-c9lbh9) ([raw file](https://github.com/cockroachdb/docs/blob/2b1d0ca373a9ab1b1b560cece11009429199e30d/v2.1/architecture/transaction-layer.md#L81)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I would simplify this to say that "for `AS OF SYSTEM TIME` queries with timestamps suitably in the past (note: we should define what "suitably" means, by default 10-20s should do), under normal operation followers will be able to ....
</blockquote></details>

Updated using that language.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3288#-:-LG74ijT1_zVPnjVwITg:b8dyj9s)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/34767d4daf55c9979994bb959f4b3fb2dcc03a6b/

tschottdorf commented:
I'm going to let this sit for a bit, but let's revisit this in a month or
so (when some of the functionality is hopefully actually present).

On Thu, Jun 28, 2018 at 4:44 PM cockroach-teamcity <notifications@github.com>
wrote:

>
> http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/34767d4daf55c9979994bb959f4b3fb2dcc03a6b/
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/docs/pull/3288#issuecomment-401167435>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135D0uyDUxnWLnDlMBe1FWuyzjEqIgks5uBUAlgaJpZM4Uv-xr>
> .
>
-- 

-- Tobias

jseldess commented:
@rmloveland, this has been pushed to 2.2, but feel free to leave this PR open until we're ready to work on it again.

jseldess commented:
@sploiselle, this is the draft follower reads documentation that I mentioned to you earlier. Since you'll be taking these docs in 2.2, we can probably close this, but please take a look and leverage this work and feel free to coordinate with @rmloveland.
